File: receiver-java/src/main/java/gt/edu/uvg/redes/receiver/HammingSecded.java
package gt.edu.uvg.redes.receiver;

public final class HammingSecded {
    private HammingSecded() {
    }

    public record Result(String dataBits, String status, int errorsDetected, int errorsCorrected) {
    }

    public static Result decode(String frameBits, int originalBitLength) {
        if (originalBitLength < 0 || originalBitLength > frameBits.length()) {
            throw new IllegalArgumentException("Longitud original inválida");
        }
        int parityBits = parityBitsFor(originalBitLength);
        int codeLength = originalBitLength + parityBits;
        int expectedLength = codeLength + 1;
        if (frameBits.length() != expectedLength) {
            throw new IllegalArgumentException("Longitud de Hamming inválida");
        }

        char[] code = frameBits.substring(0, codeLength).toCharArray();
        int syndrome = 0;
        for (int parityPosition = 1; parityPosition <= codeLength; parityPosition <<= 1) {
            int parity = 0;
            for (int position = 1; position <= codeLength; position++) {
                if ((position & parityPosition) != 0 && code[position - 1] == '1') {
                    parity ^= 1;
                }
            }
            if (parity == 1) {
                syndrome |= parityPosition;
            }
        }

        int globalParity = 0;
        for (int i = 0; i < frameBits.length(); i++) {
            if (frameBits.charAt(i) == '1') {
                globalParity ^= 1;
            }
        }

        String status = "OK";
        int detected = 0;
        int corrected = 0;
        if (syndrome != 0 && globalParity == 1) {
            if (syndrome > codeLength) {
                throw new IllegalArgumentException("Error Hamming fuera de rango");
            }
            code[syndrome - 1] = flip(code[syndrome - 1]);
            status = "CORRECTED";
            detected = 1;
            corrected = 1;
        } else if (syndrome == 0 && globalParity == 1) {
            status = "CORRECTED";
            detected = 1;
            corrected = 1;
        } else if (syndrome != 0) {
            throw new IllegalArgumentException("Hamming detectó múltiples errores");
        }

        StringBuilder data = new StringBuilder(originalBitLength);
        for (int position = 1; position <= codeLength; position++) {
            if (!isPowerOfTwo(position)) {
                data.append(code[position - 1]);
            }
        }
        return new Result(data.substring(0, originalBitLength), status, detected, corrected);
    }

    public static String encode(String dataBits) {
        int parityBits = parityBitsFor(dataBits.length());
        int codeLength = dataBits.length() + parityBits;
        char[] code = new char[codeLength];
        int dataIndex = 0;
        for (int position = 1; position <= codeLength; position++) {
            if (isPowerOfTwo(position)) {
                code[position - 1] = '0';
            } else {
                code[position - 1] = dataBits.charAt(dataIndex++);
            }
        }
        for (int parityPosition = 1; parityPosition <= codeLength; parityPosition <<= 1) {
            int parity = 0;
            for (int position = 1; position <= codeLength; position++) {
                if ((position & parityPosition) != 0 && code[position - 1] == '1') {
                    parity ^= 1;
                }
            }
            code[parityPosition - 1] = (char) ('0' + parity);
        }
        int global = 0;
        for (char bit : code) {
            if (bit == '1') {
                global ^= 1;
            }
        }
        return new String(code) + global;
    }

    public static int parityBitsFor(int dataLength) {
        int parityBits = 0;
        while (dataLength + parityBits + 1 > (1 << parityBits)) {
            parityBits++;
        }
        return parityBits;
    }

    private static boolean isPowerOfTwo(int value) {
        return (value & (value - 1)) == 0;
    }

    private static char flip(char bit) {
        return bit == '0' ? '1' : '0';
    }
}

File: sender-python/sender/hamming_secded.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Result:
    data_bits: str
    status: str
    errors_detected: int
    errors_corrected: int


def parity_bits_for(data_length: int) -> int:
    parity_bits = 0
    while data_length + parity_bits + 1 > (1 << parity_bits):
        parity_bits += 1
    return parity_bits


def _is_power_of_two(value: int) -> bool:
    return (value & (value - 1)) == 0


def encode(data_bits: str) -> str:
    parity_bits = parity_bits_for(len(data_bits))
    code_length = len(data_bits) + parity_bits
    code = [""] * code_length

    data_index = 0
    for position in range(1, code_length + 1):
        if _is_power_of_two(position):
            code[position - 1] = "0"
        else:
            code[position - 1] = data_bits[data_index]
            data_index += 1

    parity_position = 1
    while parity_position <= code_length:
        parity = 0
        for position in range(1, code_length + 1):
            if (position & parity_position) != 0 and code[position - 1] == "1":
                parity ^= 1
        code[parity_position - 1] = str(parity)
        parity_position <<= 1

    global_parity = 0
    for bit in code:
        if bit == "1":
            global_parity ^= 1

    return "".join(code) + str(global_parity)


def decode(frame_bits: str, original_bit_length: int) -> Result:
    if original_bit_length < 0 or original_bit_length > len(frame_bits):
        raise ValueError("Longitud original inválida")

    parity_bits = parity_bits_for(original_bit_length)
    code_length = original_bit_length + parity_bits
    expected_length = code_length + 1
    if len(frame_bits) != expected_length:
        raise ValueError("Longitud de Hamming inválida")

    code = list(frame_bits[:code_length])
    syndrome = 0
    parity_position = 1
    while parity_position <= code_length:
        parity = 0
        for position in range(1, code_length + 1):
            if (position & parity_position) != 0 and code[position - 1] == "1":
                parity ^= 1
        if parity == 1:
            syndrome |= parity_position
        parity_position <<= 1

    global_parity = 0
    for character in frame_bits:
        if character == "1":
            global_parity ^= 1

    status = "OK"
    detected = 0
    corrected = 0
    if syndrome != 0 and global_parity == 1:
        if syndrome > code_length:
            raise ValueError("Error Hamming fuera de rango")
        index = syndrome - 1
        code[index] = "1" if code[index] == "0" else "0"
        status = "CORRECTED"
        detected = 1
        corrected = 1
    elif syndrome == 0 and global_parity == 1:
        status = "CORRECTED"
        detected = 1
        corrected = 1
    elif syndrome != 0:
        raise ValueError("Hamming detectó múltiples errores")

    data = []
    for position in range(1, code_length + 1):
        if not _is_power_of_two(position):
            data.append(code[position - 1])

    data_bits = "".join(data)[:original_bit_length]
    return Result(data_bits, status, detected, corrected)
